python/lsst/daf/butler/core/composites.py

```diff
@@ -0,0 +1,129 @@
+        for n in ("storageClasses", "datasetTypes"):
+            for k in self[n]:
+                key = f"{n}.{k}"
+                if not isinstance(self[key], bool):
+                    raise ValueError(f"CompositesConfig: Key {key} is not a Boolean")
+
+
+class CompositesMap:
+    """Determine whether a specific datasetType or StorageClass should be
+    disassembled.
+
+    Parameters
+    ----------
+    config : `str`, `ButlerConfig`, or `CompositesConfig`
+        Configuration to control composites disassembly.
+    """
+
+    def __init__(self, config):
+        if not isinstance(config, type(self)):
+            config = CompositesConfig(config)
+        assert isinstance(config, CompositesConfig)
+        self.config = config
+
+    def doDisassembly(self, entity):
+        """Given some choices, indicate whether the entity should be
+        disassembled.
+
+        Parameters
+        ----------
+        entity : `str`, `StorageClass`, `DatasetType`
+            Thing to test against the configuration. For instance objects
+            the ``name`` attribute is used to determine a match.  A
+            `DatasetType` will first check its name, before checking its
+            `StorageClass`.  If a `str` is given, it will first be checked
+            as if it was a `DatasetType` and then as if it was a
+            `StorageClass`. If there are no matches the default will
+            be returned. If a `Datasetype` or `StorageClass` is given
+            and the `StorageClass` is not a composite, will always return
+            False.  The default for a `str` argument is always False since
+            it is not known whether the string really does refer to
+            a composite.
+
+        Returns
+        -------
+        doit : `bool`
+            Returns True if disassembly should occur; False otherwise.
+        """
+        isStr = False
+        components = None
+        datasetTypeName = None
+        storageClassName = None
+        if isinstance(entity, str):
+            datasetTypeName = entity
+            storageClassName = entity
+            isStr = True
+        elif isinstance(entity, DatasetType):
+            datasetTypeName = entity.name
+            storageClassName = entity.storageClass.name
+            components = entity.storageClass.components
+        elif isinstance(entity, StorageClass):
+            storageClassName = entity.name
+            components = entity.components
+        else:
+            raise ValueError(f"Unexpected argument: {entity:!r}")
+
+        # We know for a fact this is not a composite
+        if not isStr and not components:
+            log.debug("%s will not be disassembled (not a composite)", entity)
+            return False
+
+        matchName = "{} (via default)".format(entity)
+
+        # if we were given a str the default is always False
+        if isStr:
+            disassemble = False
+        else:
+            disassemble = self.config["default"]
```

Comment:
I don't understand why the default should be different for `str` vs. `DatasetType` or `StorageClass` instances.

Reply:
For DatasetType and Storageclass you know for a fact that the requested item is a composite and therefore the default matters. For a `str` you have no idea whether you are being given a composite and returning True for disassembly could be very dangerous.

One option is to drop `str` support completely.

Reply:
> One option is to drop str support completely.

If it doesn't make other code uglier, I'm all for it.  Seems like this is a pretty internal API so I'd lean towards simplifying it even if that reduces flexibility.